### ISSUE TITLE
[d3-fetch] Add definitions for d3-fetch (Required for D3v5)

### ISSUE THREAD

**Issue by tomwanzek** [DefinitelyTyped/DefinitelyTyped#23253]
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @tomwanzek @LeDragon @gustavderdrache 

I just looked at the release notes for the RC1 of D3 Version 5. **d3-fetch** will be added in as a replacement for **d3-request**.

So we'll need definitions for it proactively for D3 v5.

**Comment by Ledragon** [comment]
Yup, I am actually discovering that this module exists, though it has existed for a while. API surface looks small enough, so that should not be too much work.

**Comment by Ledragon** [comment]
On it

**Comment by tomwanzek** [comment]
Great. I think there might be some API similarities with **d3-request**. Since I tend to use Angular for http I'm not sure off the top of my head...

**Comment by Ledragon** [comment]
The API tends to be the same, except it uses `fetch` behind the scene, and methods return a Promise<T>... So still needs some rewrite. I take it as an opportunity to discover the API

**Comment by Ledragon** [comment]
PR #23324 submitted.
I am not happy with the `json` function implementation, but DT seems to refuse a generic on the method just to transmit it to the `Promise` return type. I think this can do as a first version, but an improvement might be required.
Also, method overloads now have to be with optional parameters and not separate methods, I got build errors on Travis with split methods...

**Comment by tomwanzek** [comment]
@Ledragon I just saw that the PR already got merged. Now that is what I call beaten to the punch 🤣 

As for the signature overload constraint restriction, in general, using the optional parameters is the recommended approach. It is more robust in terms  of passing in variations of different arguments with union types. 

In the other D3 definitions, we have often preferred the compiler-wise slightly more restrictive approach with separate signatures. The resulting TS lint error can be overruled by including the following rule into the `tslint.json` config file:

```json
{
    "extends": "dtslint/dt.json",
    "rules": {
        "unified-signatures": false,
      }
}
```

We made the choice to maintain this approach until such time as someone complained vigorously that they want to pass in a union type as an argument for a function/method where we created separate signature lines. Luckily, we have not heard much of anything so far with people complaining about resulting compiler errors and lacking flexibility.

The **main reasons** why we did  it are two-fold:

(1) For usability, there are some signature overloads, where the return type changes, based on the argument type. Unifying the signatures would not only introduce unions on the argument type, but unions on the return type side. We would create massive bloat on the consumer side, if the had to type-guard their way through return type distinctions, which are very natural to them right now. (An extreme example would be in [d3-interpolate](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/715d56ce223f468b2f977ff967c4f8c86009927e/types/d3-interpolate/index.d.ts#L40).
(2) It gives more signature-specific JSDOC comments, that while slightly duplicative when writing the file are easier to understand in Code Editors (as overloads) than a convoluted, consolidated JSDoc comment that has to explain all the separate scenarios.
